pkg/storage/azurecdn/storage.go

```diff
@@ -48,53 +48,88 @@ func (s *Storage) Save(ctx context.Context, module, version string, mod, zip, in
+	uploadOpts := func(contentType string) uploadOptions {
+		return uploadOptions{
+			accessConditions: emptyBlobAccessCond,
+			metadata:         emptyMeta,
+			blobHTTPHeaders:  httpHeaders(contentType),
+		}
+	}
 
 	const numUpload = 3
-	uploadErrs := make(chan error, numUpload)
+	errChan := make(chan error, numUpload)
 
-	upload := func(url azblob.BlockBlobURL, content io.ReadSeeker, contentType string) {
-		_, err := url.Upload(ctx, content, httpHeaders(contentType), emptyMeta, emptyBlobAccessCond)
-		uploadErrs <- err
-	}
-
-	go upload(infoBlobURL, bytes.NewReader(info), "application/json")
-	go upload(modBlobURL, bytes.NewReader(mod), "text/plain")
-	go upload(zipBlobURL, bytes.NewReader(zip), "application/octet-stream")
-
-	encountered := make([]error, 0, numUpload)
-	for i := 0; i < numUpload; i++ {
-		select {
-		case err := <-uploadErrs:
-			if err != nil {
-				encountered = append(encountered, err)
-			}
-		case <-ctx.Done():
-			return ctx.Err()
-		}
-	}
+	go uploadBlob(ctx, errChan, containerURL, module, version, "info", bytes.NewReader(info), uploadOpts("application/json"))
+	go uploadBlob(ctx, errChan, containerURL, module, version, "mod", bytes.NewReader(mod), uploadOpts("text/plain"))
+	go uploadStream(ctx, errChan, containerURL, module, version, "zip", zip, uploadOpts("application/octet-stream"))
 
-	if len(encountered) > 0 {
-		message := bytes.NewBufferString("encountered multiple errors during save:\n")
-		for _, err := range encountered {
-			fmt.Fprintln(message, err.Error())
+	var errors error
+	for i := 0; i < 3; i++ {
+		err := <-errChan
+		if err != nil {
+			errors = multierror.Append(errors, err)
 		}
-		return errors.New(message.String())
 	}
+	close(errChan)
 
 	// TODO: take out lease on the /list file and add the version to it
 	//
 	// Do that only after module source+metadata is uploaded
 
-	return nil
+	return errors
+}
+
+func blockBlobURL(containerURL azblob.ContainerURL, module, version, name string) azblob.BlockBlobURL {
+	return containerURL.NewBlockBlobURL(fmt.Sprintf("%s/@v/%s.%s", module, version, name))
+}
+
+func uploadBlob(ctx context.Context, errChan chan<- error, containerURL azblob.ContainerURL, module, version, name string, content io.ReadSeeker, opts uploadOptions) {
+	url := blockBlobURL(containerURL, module, version, name)
+	save := func() error {
+		_, err := url.Upload(ctx, content, opts.blobHTTPHeaders, opts.metadata, opts.accessConditions)
+		return err
+	}
+
+	select {
+	case errChan <- save():
+	case <-ctx.Done():
+		errChan <- fmt.Errorf("uploading %s/%s.%s timed out", module, version, name)
+	}
+}
+
+func uploadStream(ctx context.Context, errChan chan<- error, containerURL azblob.ContainerURL, module, version, name string, stream io.Reader, opts uploadOptions) {
+	bufferSize := 1 * 1024 * 1024 // Size of the rotating buffers that are used when uploading
+	maxBuffers := 3               // Number of rotating buffers that are used when uploading
```

Comment:
Get this from env/config or just sane defaults?

Reply:
I would think some sane defaults should be fine.